[WIP] Added first custom_layout code.

choldgraf opened the pull request:
Generates layouts correctly but the aligning when using a background image isn't perfect. Maybe the problem is normalizing by non-square image? Mentioned in #1758

wmvanvliet commented:
please be aware of the _box_size function:
https://github.com/mne-tools/mne-python/blob/master/mne/channels/layout.py#L419

it can automatically compute and optimal with/height for the boxes.

agramfort commented:
also you'll need to add a test

choldgraf commented:
Thanks for the tip Marijn, I'll look into that. And tests will be
forthcoming, just wanted to get the code up there first.

On Mon, Jan 26, 2015 at 12:32 AM, Alexandre Gramfort <
notifications@github.com> wrote:

> also you'll need to add a test
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-71427807>.
>

choldgraf commented:
Looks like the _box_size function works great - thanks for the tip. I'm having trouble getting an image to plot in the background correctly. Right now I'm just normalizing the xy points by the dimensions of an image, then calling plot_topo, then plotting the image underneath those plots. E.g.:

```
f = mne.viz.plot_topo(evoked, layout=layout, layout_scale=1)
for ax in f.axes:
    ax.patch.set_alpha(.1)
    
ax = plt.axes([0, 0, 1, 1])
im = ax.imshow(img)  
ax.set_zorder(-1)
```

However, it seems like there's always some blank space around the topo plots, like this:

![image](https://cloud.githubusercontent.com/assets/1839645/5929682/e84a688e-a63a-11e4-819a-5480b414584c.png)

Any ideas why it would be doing this? Ideally the entire background would be taken up by the image. I'm not sure why there's the padding at the top/bottom...is the figure window forced to be square?

Eric89GXL commented:
Looks like an aspect ratio issue to me, and this confirms it as a likely suspect:
```
import numpy as np
import matplotlib.pyplot as plt
plt.ion()
img = np.random.rand(10, 20, 3)
ax = plt.axes([0, 0, 1, 1])
im = ax.imshow(img)
ax.set_zorder(-1)
```
![figure_1](https://cloud.githubusercontent.com/assets/2365790/5929766/e8d28de4-a63b-11e4-8215-5fbd13de66b4.png)

Changing the `imshow` to have `aspect='auto'`:

![figure_2](https://cloud.githubusercontent.com/assets/2365790/5929770/f03997bc-a63b-11e4-98d1-cb6bdbc84a48.png)

Eric89GXL commented:
BTW I got the idea from here, second result for googling "matplotlib imshow autoscale":

http://stackoverflow.com/questions/13384653/imshow-extent-and-aspect

choldgraf commented:
Ah, here is the same code with aspect='auto'

![image](https://cloud.githubusercontent.com/assets/1839645/5929945/3566e9b4-a63e-11e4-88ea-3307dc5e1605.png)

How would you imagine making a test for this kind of thing? It ultimately comes down to "do the axis points seem to lie on the correct image points", but comparing the actual plotted image seems not possible in the test.

I was thinking that you could define an image with circles on it, then define the XY positions of those circles, then use those positions and that image to create a layout. Finally, see if when you scale the layout.pos values back up to their original dimensions, the xy points still fall inside of a circle.

Eric89GXL commented:
You mean for test-suite purposes, or development? I think it's probably overkill for the test suite, eyeball test seem sufficient. But if you really want to make a unit test for it, I won't stop you :)

For development, I'd do whatever was easiest to validate.

choldgraf commented:
I was thinking of test-suite stuff, but you're right, that'd be hard and
probably overkill. Is there a testing folder that you guys would use for an
"eyeball" test? AKA something that you wouldn't want nose to deal with, but
that you could refer to if anything was looking wonky?

On Tue, Jan 27, 2015 at 5:12 PM, Eric Larson <notifications@github.com>
wrote:

> You mean for test-suite purposes, or development? I think it's probably
> overkill for the test suite, eyeball test seem sufficient. But if you
> really want to make a unit test for it, I won't stop you :)
>
> For development, I'd do whatever was easiest to validate.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-71763280>.
>

Eric89GXL commented:
We usually use examples for that sort of thing. It would be nice to have a simple example that showed the functionality and why it's useful, while also serving as a check to make sure things didn't break -- many of our examples work this way.

choldgraf commented:
Sounds good - perhaps I can write an example rather than just hosting the
stuff on my blog/github account :P

On Tue, Jan 27, 2015 at 6:33 PM, Eric Larson <notifications@github.com>
wrote:

> We usually use examples for that sort of thing. It would be nice to have a
> simple example that showed the functionality and why it's useful, while
> also serving as a check to make sure things didn't break -- many of our
> examples work this way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-71770913>.
>

wmvanvliet commented:
Once the `generate_2d_layout` function is complete, the `make_eeg_layout` function should be modified to use it. DRY and all that.

agramfort commented:
travis is not happy certainly because docstring is not up to date

agramfort commented:
can you fix the travis issue?

choldgraf commented:
Yes - got a bit carried away with work the last few weeks. I will work on finishing this up in the next day or two. I'm not sure what you meant by the docstring not being up to date - I tried to model the docstring for this function in the same manner that the other functions use, but perhaps I am missing something?

agramfort commented on mne/channels/layout.py:
```diff
@@ -23,6 +23,7 @@
 from ..viz import plot_montage
 from ..transforms import (_sphere_to_cartesian, _polar_to_cartesian,
                           _cartesian_to_sphere)
+from matplotlib.pyplot import imread
```
matplotlib is not a strong dependency of MNE. You should use a nested import.

agramfort commented on mne/channels/layout.py:
```diff
@@ -931,3 +932,80 @@ def apply_montage(info, montage):
+        The image over which sensor axes will be plotted. Either a path to an
+        image file, or an array that can be plotted with plt.imshow. If
+        provided, xy points will be normalized by the width/height of this
+        image. If not, xy points will be normalized by their own min/max.
```
here docstring is not up to date.

choldgraf replied:
Ahh - thank you for the clarification. I moved the import and updated the docstring.

Eric89GXL replied:
can't this be `| None`?

agramfort commented on mne/channels/layout.py:
```diff
@@ -931,3 +931,98 @@ def apply_montage(info, montage):
+        # Pull xy coordinates interactively
+        # Can add this later but the function should work with pre-suppleid
+        # xy points first
+        pass
```
don't write code that "may" be useful in the future. Or let's keep this PR as WIP and merge when you're happy about the new feature

choldgraf commented:
Good point - I removed the unimplemented "interactive" feature. I don't think it's necessary for a working version of this function. BTW, has anyone used to this to make sure that it actually works?

Eric89GXL commented on mne/channels/layout.py:
```diff
@@ -931,3 +931,85 @@ def apply_montage(info, montage):
+
+    name : string
+        The name of this layout type.
+
```
We don't put newlines between parameter names

Eric89GXL commented:
Would it be worth adding an example of using this? That might make people more likely to test it. Also, please add some test code.

choldgraf commented:
I can write up a blog post that shows a use case for ecog. Would you prefer that it be a self-contained python script like you guys have on the website? For tests, I will try to come up with some in the coming days.

Eric89GXL commented:
Usually an example in the `examples/` directory is best, but maybe for this use case a blog post would be fine.

choldgraf commented:
Cool - I'll put something together when the tests are done, and we can
figure out the best format for it to be in. Personally I feel like
notebook-driven examples are a lot clearer for me to understand, which is
why I use them rather than writing out python scripts. However, I totally
understand if it'd be better for the project to keep things consistent and
in the "examples" folder.

Chris

On Tue, Feb 24, 2015 at 5:53 PM, Eric Larson <notifications@github.com>
wrote:

> Usually an example in the examples/ directory is best, but maybe for this
> use case a blog post would be fine.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-75890675>.
>

Eric89GXL commented:
@choldgraf I feel like this PR is pretty close. Since you're in hacking mode, maybe now would be a good time to finish? :)

choldgraf commented:
Yo - yeah this is on my list of stuff to work on next. The last few weeks
I've been in NRSA hell (just submitted on Wednesday), so I'll have more
time to work on non mind-numbing grant stuff next :)

On Fri, Apr 10, 2015 at 1:11 PM, Eric Larson <notifications@github.com>
wrote:

> @choldgraf <https://github.com/choldgraf> I feel like this PR is pretty
> close. Since you're in hacking mode, maybe now would be a good time to
> finish? :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-91670220>.
>

choldgraf commented:
Hmm - I just rebased onto master and it looks like some of the time-gen stuff got changed...what's up with that? I don't remember making any of those edits so I assume that I did something wrong on the rebase...

choldgraf commented:
As for what's left - it seems that right now the main thing to do is basically test this out under a few cases to see if it works. Here's what I'm thinking in terms of tests to ensure:

1. Generate a square of plots and make sure that the resulting x/y positions of each subplot are correct
1. Generate a custom shape of plots and make sure that x/y positions are correct
1. For a tutorial or blog post - Pull an image from the web (maybe hosted on the mne website?) and define positions on this, make sure that the plots sync up with those points.

This brings up a question - right now there's extra code in there for clicking image positions. Is this something that we want to include in the scope of MNE, or should we just leave it up to the user to figure out how to get a set of 2-d points from an image?

Eric89GXL commented:
did you `git fetch upstream` and do `git rebase upstream/master`?

choldgraf commented:
basically - there was one extra step in there:

```
git checkout master
git fetch upstream master
git merge upstream/master
git checkout this_branch
git rebase master
```

The only merge conflicts on the rebase were a couple of import differences. The time-gen stuff didn't show up there at all...

Eric89GXL commented:
I have never seen a merge and rebase play well together. Why did you do a merge, then a rebase?

Eric89GXL commented:
...and importantly, did you make a backup of your branch before doing any of this?

Eric89GXL commented:
if not, then I'd just do:
```
git checkout -b fix-rebase upstream/master
git cherry-pick 7b1611f a732fab 6fa6cc0 47ae4e5 f604d61 0ce0811 f4f9170
<use git mergetool and git cherry-pick --continue until we're good>
git push origin --force fix-rebase:add_custom_topo_layout
```

choldgraf commented:
The short answer is because I don't have a ton of experience doing rebasing and didn't realize it was an issue.

The long answer is that I figured it would be good to have my local master be up-to-date anyway, which is why I merged first and then just rebased onto my own repo's master branch.

And I'm pretty sure that I did not make a backup of this branch. Not really sure how that works...do you use a `git stash` or something? Again, not super experienced with rebasing so I'm probably being stupid.

Is the main concern the time-gen stuff? Other than that it seems like everything else looks normal. Is there a way to just remove the time-gen changes from this branch in general?

Eric89GXL commented:
Your plan sounds good. In terms of clicking, I don't have a strong opinion, but keep in mind that any code you add should be robust, broadly useful, and you will be expected to some extent to help support it going forward :)

Eric89GXL commented:
The cherry-picking line above should do what you want, it will make a fresh branch based off of upstream/master and then put your N commits on top of it only.

Not knowing the state of your master branch and so forth, I'm not sure what was wrong. My guess is that it was not a clean branch. I actually usually do this to ensure my local `master` is correct:
```
git fetch upstream
git checkout master
git reset --hard upstream/master
```
But I only do that if I actually want to work with current master. If I have a given branch I want to rebase, this is safest (doesn't depend on local `master` at all):
```
<on branch that needs to be rebased>
git fetch upstream
git rebase upstream/master
```

choldgraf commented:
OK I'll give that a shot - does that mean I'd need to open up a new pull request from the new branch?

Eric89GXL commented:
no, the last line I pasted earlier push-forces your local branch onto the same remote branch that is for this PR:
```
git push origin --force fix-rebase:add_custom_topo_layout
```

choldgraf commented:
Ah ok - I will try this in a little bit and hopefully will not explode github in the process...

Eric89GXL commented:
BTW in the future if you're new to rebasing I'd do this to backup your branch locally first:
```
git checkout branch-needing-rebase
git checkout -b branch-needing-rebase-bak
git checkout branch-needing-rebase
```
Or if you want you can always back it up remotely:
```
git push origin branch-needing-rebase:branch-needing-rebase-bak
```

choldgraf commented:
Ahh I see - that makes sense / is a good idea. I should have listened to my
general principle of always backing up anything you do on a computer before
trying something for the first time :(

On Mon, Apr 13, 2015 at 3:21 PM, Eric Larson <notifications@github.com>
wrote:

> BTW in the future if you're new to rebasing I'd do this to backup your
> branch locally first:
>
> git checkout branch-needing-rebase
> git checkout -b branch-needing-rebase-bak
> git checkout branch-needing-rebase
>
> Or if you want you can always back it up remotely:
>
> git push origin branch-needing-rebase:branch-needing-rebase-bak
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-92516431>.
>

choldgraf commented:
OK - I think that the cherry picking worked correctly, as there are now all commits are from me and only on the relevant files. I'll get back to adding functionality rather than un-breaking things next :)  thanks for the help!

agramfort commented on mne/channels/layout.py:
```diff
@@ -26,6 +26,7 @@
 from ..transforms import (_sphere_to_cartesian, _polar_to_cartesian,
                           _cartesian_to_sphere, apply_trans,
                           get_ras_to_neuromag_trans)
+from matplotlib.pyplot import imread
```
use imread from scipy.ndimage

agramfort commented on mne/channels/layout.py:
```diff
@@ -971,3 +972,80 @@ def apply_montage(info, montage):
+        functions and methods.
+    '''
+
+    from matplotlib.pyplot import imread
```
?

agramfort commented on mne/viz/utils.py:
```diff
@@ -386,3 +386,67 @@ def figure_nobar(*args, **kwargs):
+import Image
+from ImageTk import PhotoImage
+import numpy as np
+import pandas as pd
```
??? pandas is not a required dependency of mne...

choldgraf commented:
Thanks for the comments @agramfort, a lot of those were carryovers from the last round of edits (e.g., before I knew not to include pandas as an import). I've updated the clicking code to use native matplotlib rather than TK, which should make it a lot easier to maintain and understand. I think it uses the same functionality, but will try testing it out next in order to make sure the code works as it should.

agramfort commented:
ok. Can you put WIP in PR title and switch it to MRG when ready to review.

thx

choldgraf commented:
Will do

choldgraf commented:
Hey all - in terms of some kind of demo / guide, how is something like this:

http://nbviewer.ipython.org/github/choldgraf/write-ups/blob/master/mne/mne_clickable_image.ipynb

It's a bit hard to write this into an example because of the interactive clicking part, but I could create a csv file with pre-defined positions, as well as an image file, and generate the image in this notebook with it.

Eric89GXL commented:
Notes:

* You can use `python setup.py develop` to make it so if you do `import mne`, it imports whatever you have checkout out currently using `git`.

* I, too, had to do some silly iteration with `str()` to use `mne.create_info` for prototyping. Can you add integer support to `create_info`, which just creates channels named 0->N if N is passed as the channel names parameter?

* I like your example. Are you okay adding the layout and image to the mne-python repo? If so, then you could just have the code you used to generate the positions, with instructions, in a comment at the top of the example, and the rest of the example could proceed as it does in your notebook.

Sound good?

Eric89GXL commented:
(we should have functions for saving and loading layout files somewhere, so you can just use those directly instead of CSV)

Eric89GXL commented:
...and I leave it as a challenge to make the image as small as possible (JPEG with huge compression, GIF with a 2, 4, or 8 color pallette, etc.) without degrading the quality too much. :) But seriously, we don't want to add many binaries to the repo, so please make it small.

choldgraf commented:
haha - ok, I'll take a whack at this tomorrow.

On Wed, Apr 15, 2015 at 3:28 PM, Eric Larson <notifications@github.com>
wrote:

> ...and I leave it as a challenge to make the image as small as possible
> (JPEG with huge compression, GIF with a 2, 4, or 8 color pallette, etc.)
> without degrading the quality too much. :) But seriously, we don't want to
> add many binaries to the repo, so please make it small.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-93587032>.
>

choldgraf commented:
OK here are some updated:

* I fixed up the create_info code, been wanting to do that for a while.
* I got the image size down to 10kb using a GIF with 4 colors. Think that's OK?
* I'm trying to save the layout file with `my_layout.save`, but I'm getting this error:

  ```python
TypeError                                 Traceback (most recent call last)
<ipython-input-31-758106980398> in <module>()
----> 1 lt.save(layout_path + 'custom_layout.lout')

/Users/choldgraf/github/mne-python/mne/channels/layout.pyc in save(self, fname)
     69         height = self.pos[:, 3]
     70         if fname.endswith('.lout'):
---> 71             out_str = '%8.2f %8.2f %8.2f %8.2f\n' % self.box
     72         elif fname.endswith('.lay'):
     73             out_str = ''

TypeError: not enough arguments for format string
```
  `box` is a tuple of two numbers...so I think that's where the issue is coming in. Is this a bug with my code, or a bug with layout.save?

Lemme know!

Eric89GXL commented:
IIRC the box is x, y, w, h, I'm surprised things work if you only have two of them. Maybe that's a difference between .lay and .lout? I don't remember what those are, I didn't know we had two...

Eric89GXL commented:
I think 10k is okay

Eric89GXL commented on mne/channels/layout.py:
```diff
@@ -971,3 +972,76 @@ def apply_montage(info, montage):
+        ch_names = ['{0}'.format(i) for i in ch_indices]
+
+    assert len(ch_names) == len(ch_indices), 'Unequal # ch names and indices'
+    assert len(ch_names) == len(xy), 'Unequal # ch names and xy vals'
```
`assert` statements are skipped when code is compiled with optimization on, so please change these to proper checks

choldgraf replied:
By this you mean use `assert_equal` and such, correct?

Eric89GXL replied:
no those are only used in testing, in the main code we always follow this form:
```
if bad_condition:
    raise AppropriateError('you did something bad')
```

choldgraf replied:
Ahh I see - thanks for the clarification

On Thu, Apr 16, 2015 at 11:25 AM, Eric Larson <notifications@github.com>
wrote:

> In mne/channels/layout.py
> <https://github.com/mne-tools/mne-python/pull/1762#discussion_r28536215>:
>
> > +        image. If not, xy points will be normalized by their own min/max.
> > +
> > +    Returns
> > +    --------
> > +    layout : Layout
> > +        A Layout object that can be plotted with plot_topo
> > +        functions and methods.
> > +    '''
> > +
> > +    if ch_indices is None:
> > +        ch_indices = np.arange(xy.shape[0])
> > +    if ch_names is None:
> > +        ch_names = ['{0}'.format(i) for i in ch_indices]
> > +
> > +    assert len(ch_names) == len(ch_indices), 'Unequal # ch names and indices'
> > +    assert len(ch_names) == len(xy), 'Unequal # ch names and xy vals'
>
> no those are only used in testing, in the main code we always follow this
> form:
>
> if bad_condition:
>     raise AppropriateError('you did something bad')
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762/files#r28536215>.
>

choldgraf commented:
Hrm...I'm making the box with `_box_size`, which was there before my time. This function returns two floats, one for width, one for height. Maybe I'm supposed to create a full box from that?

Either way, I've updated the notebook and code to reflect recent changes. It works fine with `.lay`

Eric89GXL commented:
I'm not sure how it's supposed to work, so I can't really help you sort it out. If `lay` works, I'm okay with that, though.

Eric89GXL commented:
Travis is unhappy

choldgraf commented:
I'm gonna speak to him this morning and see why he's complaining...
On Apr 17, 2015 9:04 AM, "Eric Larson" <notifications@github.com> wrote:

> Travis is unhappy
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-94016723>.
>

choldgraf commented:
Not related to Travis, but I'm getting some weird behavior with layout saving / loading. I can create the Layout fine, but when I save, then load it back in, the `box` attribute is `None`. That said, it still seems to plot OK, which is weird...

I'm still not sure why `.lout` and `.lay` both exist separately. The functions for reading them in are a little bit different (and strangely the `.lay` reading function just sets box to `None`, which explains why this is happening). Are these for compatibility with other programs or something?

Eric89GXL commented:
Yes, I think so. I think .lout is the one that ships with mne, so it would be better to use that if you can. Can you change your code to use that? I think you'd just need to calculate a couple of extra parameters...?

choldgraf commented:
Yeah, that's what I'm working on now. I'm creating the width and height parameters with _box_size, then setting box to `(0, 0, w, h)`...we'll see if that works

Eric89GXL commented:
Cool, and your travis error is just style. You can always check at your end by running `make flake` with an up-to-date pep8, pyflakes, and flake8 installed.

Eric89GXL commented:
@choldgraf will you have time to work on this today or tomorrow at all? If not, we'll probably move it to 0.10 instead of 0.9 since we're pushing for release.

choldgraf commented:
I'll work on this a bit today - I don't think there's much left to do. The travis thing is just an EOF white space issue. Other than this I just need to include tests unless this PR needs to include things like modifying the "make eeg layout" function to use this etc.

Eric89GXL commented:
Nah, that can be done later. But yes please add tests to cover as much code as possible.

choldgraf commented:
Just added a few tests for the generate_2d_layout function. It'll be hard to write a test for the ClickableImage class, since that requires some interactivity. I've got the ipython notebook referenced above. I'm happy to turn that into a blog post for you guys, or just post it on my website or something like that. What do you think?

Also - while Travis was happy w/ the last commit, the latest appveyor build broke. Here's the traceback...maybe it's not related to this change?

```python
======================================================================
FAIL: Test that module imports are necessary
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Python27-conda64\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\projects\mne-python\mne\tests\test_import_nesting.py", line 43, in test_module_nesting
    raise AssertionError(stdout)
AssertionError: Found un-nested scipy submodules:
set(['integrate', 'ndimage', 'special'])
>>  raise AssertionError("Found un-nested scipy submodules:\r\nset(['integrate', 'ndimage', 'special'])")
```

Eric89GXL commented:
To save import speed, you need to nest any import to scipy.those modules. I'll find it.

Eric89GXL commented on mne/channels/layout.py:
```diff
@@ -26,6 +26,7 @@
 from ..transforms import (_sphere_to_cartesian, _polar_to_cartesian,
                           _cartesian_to_sphere, apply_trans,
                           get_ras_to_neuromag_trans)
+from scipy.ndimage import imread
```
nest this

Eric89GXL commented:
Then you should be good.

Eric89GXL commented:
And the clickable thing you don't absolutely have to write tests for, but if you want to write a simple test that generates a layout with a single fake click, you can. Look at `mne/viz/tests/test_raw.py`, and use the `_fake_click` to fake a click in your axes, see if it generates an appropriate plot. Sound doable? You might need to move `_fake_click` to e.g. `mne/viz/utils.py` since it'll be used in multiple places now.

choldgraf commented:
OK - I added a test using the `_fake_click` function. Thanks for the tip. Let me know how that looks!

(the failed commit was because I got OCD and updated the docstring of the test above the one I wrote so that my linter would stop yelling at me, but apparently that's been changed in master or something so I switched it back)

Eric89GXL commented:
Made a couple of minor comments.

Can you add a simple example file, that says how you would have made a useable plot from your brain image? For example, you probably need to put in a `ClickableImage` line, but comment it out, saying the result was saved to `mne/data/whatever`, then load that result, and continue (probably plotting noise inserted into a `RawArray` in this case). WDYT?

choldgraf commented:
OK, how about that?

choldgraf commented:
Oops, just saw your last comment. Yeah I can do this...it's basically what the ipython notebook was doing before. I can just scriptify that and put it in examples?

Eric89GXL commented:
No IPython, yeah just turn it into a script, also Travis is angry :)

choldgraf commented:
OK, I added an example...let me know what you think.

Also, I'm not sure why Travis is angry. It's working on all the builds except for the one with python 3.4. Here's the error message. I wonder, does 3.4 handle tuples differently or something? Or does numpy handle `shape` differently?

```python
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/testenv/lib/python3.4/site-packages/mne-0.9.dev0-py3.4.egg/mne/channels/tests/test_layout.py", line 365, in test_generate_2d_layout
    lt = generate_2d_layout(xy, w=w, h=h)
  File "/home/travis/miniconda/envs/testenv/lib/python3.4/site-packages/mne-0.9.dev0-py3.4.egg/mne/channels/layout.py", line 1027, in generate_2d_layout
    ch_indices = np.arange(xy.shape[0])
IndexError: tuple index out of range
```

Eric89GXL commented:
The example appears to displease Travis:

https://travis-ci.org/mne-tools/mne-python/jobs/59963817#L837

Py3k doesn't do anything differently with tuples, but things like `zip` are now iterators and not a list. that means that this line:
```
xy = np.array(zip(*[i.ravel() for i in square]))
```
Produces this:
```
>>> snobg = 10
>>> sbg = 15
>>> square = np.meshgrid(*[np.arange(snobg)] * 2, indexing='xy')
>>> xy = np.array(zip(*[i.ravel() for i in square]))
>>> xy
array(<zip object at 0x101fd03c8>, dtype=object)
```
Notice that this is a array with one element, which will have `ndim=0` and thus `len(shape) == 0`.

choldgraf commented:
Ohhhhh that's it. OK, so I need to do `list(zip(...` and it should be OK
then, right?

I should really get more familiar with python 3 :-/

On Fri, Apr 24, 2015 at 7:31 PM, Eric Larson <notifications@github.com>
wrote:

> The example appears to displease Travis:
>
> https://travis-ci.org/mne-tools/mne-python/jobs/59963817#L837
>
> Py3k doesn't do anything differently with tuples, but things like zip are
> now iterators and not a list. that means that this line:
>
> xy = np.array(zip(*[i.ravel() for i in square]))
>
> Produces this:
>
> >>> snobg = 10
> >>> sbg = 15
> >>> square = np.meshgrid(*[np.arange(snobg)] * 2, indexing='xy')
> >>> xy = np.array(zip(*[i.ravel() for i in square]))
> >>> xy
> array(<zip object at 0x101fd03c8>, dtype=object)
>
> Notice that this is a array with one element, which will have ndim=0 and
> thus len(shape) == 0.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-96122201>.
>

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,65 @@
+
+# Set parameters and paths
+plt.rcParams['image.cmap'] = 'gray'
+im_path = '/Users/choldgraf/github/mne-python/mne/data/image/mni_brain.gif'
```
I can't run this example currently -- this will need to be more like `op.path(op.dirname(__file__), '..', 'mne', 'data', 'image', 'mni_brain.gif')`

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,65 @@
+im_path = '/Users/choldgraf/github/mne-python/mne/data/image/mni_brain.gif'
+
+# We've already clicked and exported
+layout_path = '/Users/choldgraf/github/mne-python/mne/data/image/custom_layout.lout'
```
same here

Eric89GXL commented:
I would prefer to find a more explicit way to do what you're trying to accomplish. I love one-liners, but an ndarray cast of a list of a zip of a `*` of a list comprehension is a bit opaque even for me :)

choldgraf commented:
Haha, I've been using pandas too much. OK, I'll try to do something more
elegant. Maybe just use itertools rather than meshgrid?

I will try to finish this up either tonight or tomorrow!

On Fri, Apr 24, 2015 at 7:45 PM, Eric Larson <notifications@github.com>
wrote:

> I would prefer to find a more explicit way to do what you're trying to
> accomplish. I love one-liners, but an ndarray cast of a list of a zip of a
> * of a list comprehension is a bit opaque even for me :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-96122610>.
>

Eric89GXL commented:
Do you need to use itertools? Could you use a `for` loop? The farther you wander from builtins and `numpy` generally the less readable code becomes. `meshgrid` is fine if you need to use it, most people will know what it does.

Eric89GXL commented:
IIRC some things from itertools have moved, too, so you might end up needing to get them from somewhere in `externals.six`.

choldgraf commented:
Ok ok, I will try it in raw python, I promise :-)
On Apr 24, 2015 7:50 PM, "Eric Larson" <notifications@github.com> wrote:

> IIRC some things from itertools have moved, too, so you might end up
> needing to get them from somewhere in externals.six.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-96122768>.
>

choldgraf commented:
OK - I think that I've addressed both of these things (still managed to keep the xy definition a one-liner...think that's OK as is?). It looks like appveyor is crashing but I'm having a hard time getting information about it. I'll check again later tonight or tomorrow.

Eric89GXL commented:
All good from Travis except for style:

https://travis-ci.org/mne-tools/mne-python/jobs/59979483#L841

It might we worth getting automated PEP8 / pyflakes checking. If you don't want to do that, you can install `flake8` and do `make flake` before pushing, it would save ~1.5 hours of CI run time if you expect frequent style errors.

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,67 @@
+from mne.channels.layout import generate_2d_layout
+from os import path as op
+
+# Set parameters and paths
```
Please put `print(__doc__)` right after the imports (see other examples)

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,67 @@
+epochs = mne.EpochsArray(data, info, events)
+
+# Using the native plot_topo function
+f = mne.viz.plot_topo(epochs.average(), layout=lt)
```
Maybe instead of showing it with and without, instead show the image by itself up by the commented-out code (without calling `show`) and then just show the full thing, as you currently do, below. It's easy enough to imagine what the traces would look like without a background image. But it would be nice to see the image without traces on it so people could see that it had "target" click locations on it.

Eric89GXL commented:
Two remaining issues:

First, when I click inside the plot but not on a trace, I get this:
```
  File "/Users/larsoner/custombuilds/mne-python/mne/viz/topo.py", line 184, in _plot_topo_onpick
    raise err
AttributeError: 'Axes' object has no attribute '_mne_ch_idx'
```
Second, if I resize the figure, the alignment gets screwed up. You either need to make it so that the image is *not* fixed aspect ratio (which isn't great), or you need to modify the `matplotlib` axes placement so that they follow a fixed aspect ratio. This is probably the correct solution, but I expect it will be much harder to figure out, so if you want to go with the non-fixed version it's okay.

If you want to go for the better solution, Google might pop up something from StackOverflow or the matplotlib mailing list. You could also just shoot a quick email to the mailing list with your use case (placing axes on specific locations on a fixed-aspect-ratio background image and having them stay in the proper positions regardless of window size) and see if someone responds. There must be a way to change the axis placement during resize... then it's a matter of dealing with the transformation systems to get the transforms right. Maybe on each resize, you could take the original locations in data (image x,y) coordinates, transform them out to figure coordinates, and then relocating each axis appropriately. That actually sounds doable.

Eric89GXL commented:
Actually, it looks like you *might* just be able to put everything into an `AxesDivider` and set the aspect ratio of the divider to 1 as well:

http://matplotlib.org/1.4.2/mpl_toolkits/axes_grid/users/axes_divider.html

I'd try that first. :)

choldgraf commented:
Thanks for your comments - I've fixed the PEP issue (I actually have PEP8/257 modules in sublime text, I just missed this one). I didn't think about the fact that building these PRs is expensive/wasteful. Sorry about that - I'll do more QA before pushing new commits next time.

Regarding the resizing / clicking errors, I totally didn't think about that. I will try to address these tomorrow. I've just been using this code on images of fixed size, and just kept everything consistent from there (if I wanted an image that was a different size / a crop / etc, then I just photoshopped the image and then re-clicked)

Eric89GXL commented:
The build time isn't a huge deal, just something to keep in mind :)

Yeah just make sure it works well in resizing, which I think should work of
using the divider class and also setting its aspect ratio. But maybe you
can't also custom position axes, in which case you'll have to set positions
manually on resize. Or you could allow your image to deform on resize I
guess...

choldgraf commented:
Hey @Eric89GXL, looking into your other two comments now, and I'm a bit confused as to the errors you're getting. Questions below...

Two remaining issues:

>First, when I click inside the plot but not on a trace, I get this:
```python
  File "/Users/larsoner/custombuilds/mne-python/mne/viz/topo.py", line 184, in _plot_topo_onpick
    raise err
AttributeError: 'Axes' object has no attribute '_mne_ch_idx'
```python

What do you mean by "inside the plot but not on a trace"? This seems like it is unrelated to the additional code in this PR, but please correct me if I'm wrong. I didn't change anything in topo.py, but is this happening because of some information that I'm not including in the layout file or something?

I tried clicking outside of the displayed image in ClickableImage, and it seems fine. Nothing happens and nothing is appended to the `coords` attribute, which seems OK to me. If this is an issue with the `plot_topo` functionality, perhaps I can give it a stab but I am less familiar with that code.

> Second, if I resize the figure, the alignment gets screwed up. You either need to make it so that the image is not fixed aspect ratio (which isn't great), or you need to modify the matplotlib axes placement so that they follow a fixed aspect ratio. This is probably the correct solution, but I expect it will be much harder to figure out, so if you want to go with the non-fixed version it's okay.

What do you mean by the alignment gets screwed up? I tried resizing the image to a bunch of different ratios and clicking the same point, and it always returned the same value. Doesn't this seem like the behavior that we would want? E.g., here's the output of `plot_clicks` for 3 clicks on the same electrode with a different wonky resize each time (e.g., scaling and stretching)

![image](https://cloud.githubusercontent.com/assets/1839645/7339635/e0d588a0-ec2a-11e4-88d8-2bfce893f861.png)

Eric89GXL commented:
The stretching issue comes up not when using ClickableImage, but when actually using the `plot_topo` with a background image, i.e. your example file. If you run that example and resize the window, the plots are no longer in the correct place, right?

If you get the error about channel indices in other examples, too, please open an issue. You shouldn't have to fix it as part of this PR but one of us should fix it before release. :)

Eric89GXL commented:
For example:

![figure_2](https://cloud.githubusercontent.com/assets/2365790/7339715/cc11daf6-ec2d-11e4-8f2a-8306ba525c11.png)

choldgraf commented:
Ahhh, yes I see. So both of these comments are w/r/t the plot_topo function, not the `ClickableImage` class. That does sound like an issue if we're going to have people plot things on top of background images...

And you don't like the idea of just allowing the topo plots to have a variable aspect ratio? I was thinking: since the background plotting is already a little bit unintuitive (e.g., needing to change the zorder and such), then it might be helpful to just include another plotting function `add_topo_bg_image` or something.

This would take the topo figure and the image array as an input, then call imshow, set the zorder, and set the aspect ratio of both axes to "auto" so that they can scale normally. Something like this:

```python
def add_topo_bg_image(f, im):
    f.axes[0].set_aspect('auto')
    ax_im = f.add_axes([0, 0, 1, 1])
    ax_im.imshow(im, aspect='auto')
    ax_im.set_zorder(-1)
    return f

# And here's the call
f = mne.viz.plot_topo(epochs.average(), layout=lt)
add_topo_bg_image(f, im)
plt.show()
```
Then you could resize, e.g.:
![image](https://cloud.githubusercontent.com/assets/1839645/7339792/d9cc0912-ec2f-11e4-8dbd-3737583fae22.png)

Eric89GXL commented:
Yeah, lets just let the aspect ratio be auto. It might look like crap for some images, but in that case users can manually set the figure size correctly. It will make the coding a lot easier.

I'm fine with `mne.viz.add_background_image(fig, img)`, it should work for any figure and any image so no reason put the `topo` name in there. It will only be a few lines but worth it for usability I think.

choldgraf commented:
OK - take a look at that and let me know what you think. I also noticed that the sub-plot clicking only works when you have `qt` as the backend (at least, it doesn't work on my computer until I explicitly call `%matplotlib qt` in ipython. Should we include the `plt.set_backend` call in our examples?

Also as an aside, would it be worth including a note in the docstring of plot_topo etc. that interactive stuff may only work with qt?

Eric89GXL commented:
It might be a good idea to tell people that, yeah. I'd rather do that than explicitly set the backend.

Flake8 got you again:

https://travis-ci.org/mne-tools/mne-python/jobs/60171513#L840

BTW if there are lines you think really should be included, you can add `  # noqa` at the end of the line to get flake8 to ignore them. But this should be used very sparingly.

Example works better now. We are close!

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,70 @@
+"""
+=========================================
+Demonstration of how to use ClickableImage / generate_2d_layout.
+=========================================
```
extend to length of the title

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,70 @@
+#
+# License: BSD (3-clause)
+import sys
+sys.path.insert(0, '../')
```
please remove this. BTW if you need this because you need to import mne-python, do `python setup.py develop`. It basically symlinks your dev version of mne-python so when you do `import mne` it imports whatever you currently have checked out via `git`.

choldgraf replied:
ah whoops, missed that one. I'm going to re-build mne locally after this PR, but have been avoiding doing this until all my open PRs are finished.

Eric89GXL replied:
with `python setup.py develop` you never have to worry about re-building since we are pure python, doing `git checkout` automatically switches the active version

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,70 @@
+import mne
+from mne.viz.utils import ClickableImage, add_background_image
+from mne.channels.layout import generate_2d_layout
+from os import path as op
```
we always do general imports, then mne-python imports, so move this up

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,70 @@
+from matplotlib import pyplot as plt
+import mne
+from mne.viz.utils import ClickableImage, add_background_image
+from mne.channels.layout import generate_2d_layout
```
make it so this can be imported from `mne.channels` instead, that is too nested

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,70 @@
+                  'image', 'mni_brain.gif')
+
+# We've already clicked and exported
+layout_path = op.join(op.dirname(__file__), '..', 'mne', 'data', 'image', '')
```
you shouldn't need the trailing `''`, if you do something in the underlying code is probably wrong (i.e., not using `os.path`)...

Eric89GXL commented on mne/channels/__init__.py:
```diff
@@ -4,7 +4,7 @@
 
 from .layout import (Layout, make_eeg_layout, make_grid_layout, read_layout,
-                     find_layout, read_montage, apply_montage)
+                     find_layout, read_montage, apply_montage, generate_2d_layout)
```
the import is already here, so you can just change the line above

Eric89GXL commented:
Please also add your new functions to `doc/python_reference.rst` and update `doc/whats_new.py`.

choldgraf commented:
So the travis problem is that I'm importing `ClickableImage` and `generate_2d_layout`, but these are both in the commented-out part of the code, so travis thinks I'm never using them. Is this worthy of `# noqa`? If so, do I put this after the imports?

Eric89GXL commented:
sure, that's resonable, just put `  #noqa` at the end of the offending lines. You can do `git grep noqa` to see other examples in action in our codebase.

Eric89GXL commented:
dang it, like this:
```
line  # noqa
```

choldgraf commented:
OK - I ran `make flake` on this one before uploading, so if Travis isn't happy then I will throw my computer out the window :)

I believe that I've addressed all of the comments, let me know if there's anything else!

Eric89GXL commented on doc/source/whats_new.rst:
```diff
@@ -8,6 +8,9 @@ Current
 ~~~~~~~~~
+   - Add support for capturing cursor positions when clicking on an image by `Chris Holdgraf`_
+
+   - Add support for custom sensor positions when creating Layout objects by `Chris Holdgraf`_
```
please add to the bottom not the top (helps us keep track)

Eric89GXL replied:
bottom of this release version, that is

Eric89GXL commented on mne/channels/layout.py:
```diff
@@ -971,3 +971,82 @@ def apply_montage(info, montage):
+    --------
+    layout : Layout
+        A Layout object that can be plotted with plot_topo
+        functions and methods.
```
Please add a `Notes` section with `versionadded`, see https://github.com/mne-tools/mne-python/pull/2029/files

Eric89GXL commented on mne/channels/tests/test_layout.py:
```diff
@@ -349,3 +350,35 @@ def test_box_size():
+    assert_allclose(lt_bg.pos[:, :2].max(), xy.max() / float(sbg))
+
+if '__name__' == '__main__':
+    test_generate_2d_layout()
```
`from mne.utils import run_tests_if_main` above, then do `run_tests_if_main()` here, not inside an `if` statement

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,110 @@ def figure_nobar(*args, **kwargs):
+    im: ndarray
+        A numpy array that works with a call to
+        plt.imshow(im). This will be plotted
+        as the background of the figure.
```
`versionadded`. also, please add this to `python_reference`

Eric89GXL replied:
ahh it's in ref, never mind. still need versionadded though

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,110 @@ def figure_nobar(*args, **kwargs):
+        plt.imshow(im). This will be plotted
+        as the background of the figure.
+    """
+    fig.axes[0].set_aspect('auto')
```
What is `fig.axes[0]`, just the first axis of the figure, if it has one? Why modify that here? Also, please add a simple test of this function in your tests.

choldgraf replied:
Yes - this function is assuming that you're passing it a figure with a plot you've already created (generally an image). It sets the aspect ratio to 'auto' per our previous conversation. Were you imagining that this wouldn't happen here, but instead would happen in the `plot_topo` function? Alternatively, I could have the user pass the axis object to the function, and it would then get the figure from `axis.figure`...

Eric89GXL replied:
I'm thinking of this as a general-purpose function to add a background image to a figure. A figure in general can have any number of axes, or no axes. Thus the specialization it seems should live in the `plot_topo` function. I guess you could have an `axes` parameter here that accepts an `axes` or list of `axes` and sets them all to auto aspect ratio if need be...?

choldgraf replied:
Hmmm - what if there was a `set_aspect_ratio` flag, and if if you pass it a
parameter, then it will try:

```python
for ax in fig.axes:
  ax.set_aspect(set_aspect_ratio)
```

or something like this?

On Mon, Apr 27, 2015 at 12:09 PM, Eric Larson <notifications@github.com>
wrote:

> In mne/viz/utils.py
> <https://github.com/mne-tools/mne-python/pull/1762#discussion_r29179732>:
>
> > +    be done with topo plots, though it could work
> > +    for any plot.
> > +
> > +    Note: this sets the aspect ratio of the first
> > +    axis in "fig" to 'auto'.
> > +
> > +    INPUTS
> > +    --------
> > +    fig: plt.figure
> > +        The figure you wish to add a bg image to.
> > +    im: ndarray
> > +        A numpy array that works with a call to
> > +        plt.imshow(im). This will be plotted
> > +        as the background of the figure.
> > +    """
> > +    fig.axes[0].set_aspect('auto')
>
> I'm thinking of this as a general-purpose function to add a background
> image to a figure. A figure in general can have any number of axes, or no
> axes. Thus the specialization it seems should live in the plot_topo
> function. I guess you could have an axes parameter here that accepts an
> axes or list of axes and sets them all to auto aspect ratio if need be...?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762/files#r29179732>.
>

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,119 @@ def figure_nobar(*args, **kwargs):
+        fig.canvas.button_press_event(x, y, 1, False)
+
+
+def add_background_image(fig, im):
```
adding a `set_ratios` flag is fine by me

choldgraf replied:
Mmk, how's that? Also added a test.

On Mon, Apr 27, 2015 at 12:21 PM, Eric Larson <notifications@github.com>
wrote:

> In mne/viz/utils.py
> <https://github.com/mne-tools/mne-python/pull/1762#discussion_r29180898>:
>
> > +
> > +def _fake_click(fig, ax, point, xform='ax'):
> > +    """Helper to fake a click at a relative point within axes."""
> > +    if xform == 'ax':
> > +        x, y = ax.transAxes.transform_point(point)
> > +    elif xform == 'data':
> > +        x, y = ax.transData.transform_point(point)
> > +    else:
> > +        raise ValueError('unknown transform')
> > +    try:
> > +        fig.canvas.button_press_event(x, y, 1, False, None)
> > +    except:  # for old MPL
> > +        fig.canvas.button_press_event(x, y, 1, False)
> > +
> > +
> > +def add_background_image(fig, im):
>
> adding a set_ratios flag is fine by me
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762/files#r29180898>.
>

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,68 @@
+from matplotlib import pyplot as plt
+from os import path as op
+import mne
+from mne.viz.utils import ClickableImage, add_background_image  # noqa
```
Let's get these from `mne.viz` directly, too -- again too nested :)

Eric89GXL commented on mne/channels/layout.py:
```diff
@@ -971,3 +971,87 @@ def apply_montage(info, montage):
+
+    Notes
+    -----
+        .. versionadded:: 0.9.0
```
no need to indent this far (move back 4 spaces)

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,132 @@ def figure_nobar(*args, **kwargs):
+    Upon clicking, the x/y coordinate of the cursor will be stored in
+    self.coords, which is a list of (x, y) tuples.
+
+    Inputs
```
Parameters

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,132 @@ def figure_nobar(*args, **kwargs):
+    ------
+    imdata: ndarray
+        The image that you wish to click on for 2-d points.
+
```
add:
```
**kwargs : dict
    Keyword arguments.
```

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,132 @@ def figure_nobar(*args, **kwargs):
+
+    Notes
+    -----
+        .. versionadded:: 0.9.0
```
again no extra indent

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,132 @@ def figure_nobar(*args, **kwargs):
+        A numpy array that works with a call to
+        plt.imshow(im). This will be plotted
+        as the background of the figure.
+    set_ratios: str | an input that works with ax.set_aspect
```
If it defaults to `None`, `None` should be in the list

Eric89GXL replied:
and I think it's okay just to say it must be `str | None`, even if other params in principle could be used

Eric89GXL commented on mne/viz/utils.py:
```diff
@@ -386,3 +385,132 @@ def figure_nobar(*args, **kwargs):
+
+    Notes
+    -----
+        .. versionadded:: 0.9.0
```
no extra indent

choldgraf commented:
OK, I think these issues have been addressed but now it's non-mergeable. Think that's because of the docstring stuff? Is there a way for me to see what I changed that broke the mergeability?

choldgraf commented:
Or perhaps I need to fetch the latest master and update my own, then check for merge conflicts there?

Eric89GXL commented:
This is what I do in these situations:
```Bash
git checkout add_custom_topo_layout
git checkout -b add_custom_topo_layout-bak  # make a backup
git checkout add_custom_topo_layout
git fetch upstream
git rebase upstream/master
git mergetool
git rebase --continue
git mergetool
git rebase --continue
...
git push origin --force add_custom_topo_layout
```

choldgraf commented:
OK, I think that I fixed the rebase but let me know if it seems like I exploded anything in the process. I think it's fitting that this PR was revived, and ultimately closed, with a rebase :P is that common?

Eric89GXL commented:
Holy crap were there really 43 commits? Did I make you change things that many times???

Eric89GXL commented:
Ahh, you have some `fixing rebase` commits in there, phew :)

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,68 @@
+
+# Using the native plot_topo function with the image plotted in the background
+f = mne.viz.plot_topo(epochs.average(), layout=lt)
+ax = add_background_image(f, im)
```
I don't think this is working for me, I get a topoplot, and then in a separate figure, I get the BG image. Does it work properly for you?

choldgraf replied:
Oh yeah, that's weird. Hmmm, I guess that's related to putting the code in the function instead of hard-coding it into the script?

choldgraf replied:
Nope, this doesn't seem to happen when I put the code back into the script...

Eric89GXL replied:
try putting `show=False` in the `plot_topo` call...?

Eric89GXL replied:
works for me with that

choldgraf replied:
Yeah, that's working for me too. That is weird...do you think the rebase changed something in plot_topo?

Eric89GXL replied:
It might have a `show=True` where it didn't before...? In any case, we have the fix for you now :)

choldgraf replied:
Ya - I updated the example to show that. Too bad that this won't work
unless you put show=False...I feel like that's going to confuse people if
they use it :-/

On Mon, Apr 27, 2015 at 3:41 PM, Eric Larson <notifications@github.com>
wrote:

> In examples/plot_clickable_image.py
> <https://github.com/mne-tools/mne-python/pull/1762#discussion_r29199233>:
>
> > +lt = mne.channels.read_layout(layout_name, path=layout_path, scale=False)
> > +
> > +# Create some fake data
> > +nchans = len(lt.pos)
> > +nepochs = 50
> > +sr = 1000
> > +nsec = 5
> > +events = np.arange(nepochs).reshape([-1, 1])
> > +events = np.hstack([events, np.zeros([nepochs, 2])])
> > +data = np.random.randn(nepochs, nchans, sr * nsec)
> > +info = mne.create_info(nchans, sr, ch_types='eeg')
> > +epochs = mne.EpochsArray(data, info, events)
> > +
> > +# Using the native plot_topo function with the image plotted in the background
> > +f = mne.viz.plot_topo(epochs.average(), layout=lt)
> > +ax = add_background_image(f, im)
>
> It might have a show=True where it didn't before...? In any case, we have
> the fix for you now :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762/files#r29199233>.
>

choldgraf commented:
Haha, well, when I do commits I tend to break them up into chunks to make it more clear what I changed. E.g., if you made comments on three files that are unrelated, then I'll change all three files and then create three commits, one for each change. Maybe that's a bad idea?

Eric89GXL commented:
I pretty much always use `git commit -am 'FIX: etc.'` because what you do sounds too hard :)

Eric89GXL commented on examples/plot_clickable_image.py:
```diff
@@ -0,0 +1,68 @@
+epochs = mne.EpochsArray(data, info, events)
+
+# Using the native plot_topo function with the image plotted in the background
+f = mne.viz.plot_topo(epochs.average(), layout=lt, show=False)
```
Well, we could hijack the `fig_facecolor='k'` argument so that if a 2- or 3-D `ndarray` is supplied, it uses a background image instead of a color I guess...? Then this could just be
```
plot_topo(epochs.average(), layout=lt, fig_facecolor=im)
```
No need for `show` or anything.

Eric89GXL replied:
You would need to update the docstring and triage `fig_facecolor` appropriately. I think the way to do it would be to try to convert it to a color using a matplotlib ColorConverter (`git grep` it) and catch the appropriate exception to try it as a BG image instead.

choldgraf replied:
that would be cool - but that's not currently supported, right? It seems like then you'd want to call it "fig_background" and then explain that this could be a call to a matplotlib color, or an image...

And for what it's worth, I did just check the diff of my backup branch (aww yesss I made a backup) on upstream/master and the show=True argument was in fact added.

Eric89GXL replied:
The name `fig_facecolor` is indeed a bit of a misnomer, but not worth the deprecation headache just to get it to `fig_background` IMO

choldgraf replied:
Why not just check if it's an ndarray of len(shape) > 1, and if so treat it as an image, if not then do whatever plot_topo currently does with that argument?

Eric89GXL replied:
It can be string, tuple, list, ndarray, etc. ColorConverter is designed to convert all of these to a standard tuple, and fail with some exception if it doesn't. It's better to outsource the work here if you can.

choldgraf replied:
Why not just add an extra kwarg in there? fig_background could be something on top of fig_facecolor, since it does something different anyway. Or do you guys not want to clutter things too much?

Eric89GXL replied:
My code proposal is:
```
from matplotlib.colors import ColorConverter
color_converter = ColorConverter()
bg_img = None
try:
    fig_facecolor = color_converter.to_rgb(fig_facecolor)
except ValueError:
    img = fig_facecolor
    fig_facecolor = 'k'
```

Eric89GXL replied:
I guess you could. Maybe it is clearer...

Yeah why not, one more `kwarg`. @agramfort may complain but we can fix in another PR if need be :)

choldgraf replied:
OK, should I just go for it?

Eric89GXL replied:
Yeah, then we can hopefully merge...

choldgraf commented:
How about that?

Eric89GXL commented:
Oh, nevermind. I'm dumb :)

Eric89GXL commented:
I'll run the example and if the CIs are happy I think we can merge.

Eric89GXL commented:
Example worked great, but:
```
larsoner@bunk:~/custombuilds/mne-python$ nosetests mne/tests/test_docstring_parameters.py
F
======================================================================
FAIL: Test module docsting formatting
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/larsoner/custombuilds/mne-python/mne/tests/test_docstring_parameters.py", line 163, in test_docstring_parameters
    raise AssertionError(msg)
AssertionError: 
mne.viz.utils.ClickableImage.onclick arg mismatch: ['event']
>>  raise AssertionError("\nmne.viz.utils.ClickableImage.onclick arg mismatch: ['event']
```

choldgraf commented:
So it is a problem because the function didn't have documentation for that arg? AKA does this fix it?

choldgraf commented:
In that case should I do the same for the two `**kwargs` arguments in that class?

Eric89GXL commented:
Yep, that should do it. Try running the line I ran above to see for yourself, though.

Eric89GXL commented:
We usually do add docstring for `**kwargs : dict \n optional arguments for whatever`, too, even if the test doesn't detect it's missing

choldgraf commented:
I see - OK I think that should be covered now. Ran the docstring test you just did and it worked fine

Eric89GXL commented:
LGTM and the example still works :) Thanks again for trudging through this, but I think we've converged on a nice solution now.

choldgraf commented:
OK cool - so, we're good? Assuming travis and appveyor are appeased?

No no, thanks to you for helping me through this. Hopefully it is time well-spent and I won't have to have as much hand-holding next time around :)

Eric89GXL commented:
Yep, should be good. Fingers crossed...

Eric89GXL commented:
Thanks @choldgraf!

dengemann commented:
Really nice work @choldgraf, thanks for the contribution!

2015-04-28 4:32 GMT+02:00 Eric Larson <notifications@github.com>:

> Thanks @choldgraf <https://github.com/choldgraf>!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-96888908>.
>

choldgraf commented:
Hallelujah :D

Thanks guys!

On Mon, Apr 27, 2015 at 7:33 PM, Denis A. Engemann <notifications@github.com
> wrote:

> Really nice work @choldgraf, thanks for the contribution!
>
> 2015-04-28 4:32 GMT+02:00 Eric Larson <notifications@github.com>:
>
> > Thanks @choldgraf <https://github.com/choldgraf>!
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-96888908
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1762#issuecomment-96889022>.
>

agramfort commented:
:beers: !
